### PR TITLE
feat: add development-only local mint server

### LINKED ISSUE #1963: Standalone dev mint server (no GCP required)
## Problem

Evaluating fullsend requires provisioning GCP infrastructure (Cloud Functions, Secret Manager, WIF) before the platform can be tested. This creates a significant barrier to adoption — teams cannot try fullsend without first committing to a cloud provider.

## Solution

Add a standalone development token mint (`fullsend mint run`) that replaces GCP infrastructure with local equivalents:
- **Secret Manager → disk**: PEM files stored in a local data directory
- **WIF → JWKS**: OIDC tokens verified directly via GitHub's JWKS endpoint
- **Cloud Functions → HTTP server**: Plain HTTP server on localhost, optionally exposed via cloudflared tunnel

## Security trade-offs

- OIDC verification is enabled by default; `--insecure-no-auth` is available for local-only testing and cannot be combined with `--tunnel`
- Permissive workflow defaults (`*`) are an accepted risk for short-lived development use; `--allowed-workflows` provides tighter controls
- PEM files are stored unencrypted on disk, protected by filesystem permissions

## Acceptance criteria

- [ ] `fullsend mint run --data-dir <dir>` starts a local mint that serves real GitHub App installation tokens
- [ ] `fullsend admin install --skip-mint-check --mint-url http://localhost:8321 --mint-data-dir <dir>` writes PEMs to disk
- [ ] Optional `--tunnel` flag creates a cloudflared quick tunnel for GitHub Actions access
- [ ] OIDC verification works without GCP (direct JWKS validation)
- [ ] Documentation covers setup, security model, and limitations

### PR DESCRIPTION
**Pull request by ggallen**
## Summary

Add a standalone token mint for local development and evaluation. This eliminates the need for GCP infrastructure (Secret Manager, WIF, Cloud Functions) to get started with fullsend.

### Dev mint server (`internal/devmint/`)

- Local HTTP server that mints real GitHub App installation tokens
- Disk-based PEM storage — write PEM files and the mint picks them up on startup
- OIDC verification via JWKS (direct signature validation, no STS exchange)
- Optional `--insecure-no-auth` flag for local dev
- **cloudflared tunnel** support for exposing the local mint to GitHub Actions runners
- `/v1/status` endpoint gated behind OIDC auth (matching production mint behavior)

### CLI changes

- `fullsend mint run` command with `--data-dir`, `--port`, `--tunnel`, `--insecure-no-auth`, `--oidc-audience` flags
- `--mint-data-dir` flag on `fullsend admin install` — writes PEMs directly to disk instead of Secret Manager
- Enhanced `--mint-url` / `--skip-mint-check` validation (allows `http://localhost`, `http://127.0.0.1`, `http://[::1]` for dev mint)
- `storePEMToDisk()` for disk-based PEM persistence during install

### Documentation

- New guide: `docs/guides/infrastructure/dev-mint.md`
- Updated installation guide and docs README
- Added `mint run` to CLI command tree in `cli-internals.md`

## Review focus

1. Is `storePEMToDisk` safe against path traversal via role names?
2. Is the cloudflared tunnel lifecycle robust (startup wait, shutdown, URL extraction)?
3. Do `--insecure-no-auth` and `--mint-data-dir` compose correctly with existing flags?

## Test plan

- [x] `go test ./internal/devmint/...` — tests pass
- [x] `go test ./internal/cli/...` — CLI tests pass (including IPv6 localhost)
- [x] `go vet ./...` — clean
- [x] `go test ./...` — full test suite passes
- [ ] Manual: `fullsend mint run` → tunnel → `fullsend admin install` → agent runs

🤖 Generated with [Claude Code](https://claude.com/claude-code)

Closes #1963

### CODE REVIEW

File: docs/guides/infrastructure/dev-mint.md [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+The standalone mint is a local HTTP server that replaces the GCP token mint infrastructure (Secret Manager, Workload Identity Federation, Cloud Functions) for development and evaluation. It stores GitHub App PEM keys on disk and mints real installation tokens — no cloud infrastructure required for token minting.
+
+> **This guide is for developers and evaluators** who want to try fullsend without deploying the full GCP mint stack. For production deployments, see the infrastructure reference in the admin guides.
+
+> **Note:** The standalone mint eliminates GCP for **token minting** only. Agents still need an LLM to do their work — fullsend currently uses Claude on Vertex AI, which requires a GCP project with the Vertex AI API enabled. The `--inference-project` flag during install configures this. Without it, agents will start but fail when they try to call the model.
+
+## Prerequisites
+
+- **fullsend CLI** (v0.5.0+ or built from source):
```
Comment (rh-hemartin) [inline docs/guides/infrastructure/dev-mint.md]: This will available after 0.13.0.

**Comment by ggallen** [comment]
Rebased on main (which now includes #1783 mintcore extraction). All review findings addressed:

**Medium:**
- **[data-exposure]** `/v1/status` now requires OIDC authentication when `--insecure-no-auth` is not set, matching the Cloud Function handler's behavior.
- **[stale-doc]** Added `mint run` to the CLI command tree in `cli-internals.md`.

**Low:**
- **[error-handling-idiom]** `storePEMToDisk` now explicitly distinguishes `IsNotExist` from other read errors for `config.json`.
- **[missing-test]** Added `http://[::1]:8321` test case to `TestValidateSkipMintCheck`.
- **[edge-case]** `loadFromDisk()` now wraps shared state writes in `s.mu.Lock()`/`s.mu.Unlock()`.

Also updated mintcore API calls to match the merged #1783 interface (`JWKSVerifierConfig` struct, 4-arg `ValidateWorkflowRef`).

File: docs/guides/infrastructure/dev-mint.md (hunk 1 above)
Comment (ggallen) [reply docs/guides/infrastructure/dev-mint.md]: Updated the version requirement to v0.14.0+ to reflect the correct release target.

File: internal/cli/admin.go [hunk 2]
```diff
@@ -508,18 +587,20 @@ Inference authentication:
 
 			// Collect agent credentials via app setup.
 			var agentCreds []layers.AgentCredentials
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/cli/admin.go:622]: Fixed:

1. **ensureConfigRepoExists gate**: Now also creates the config repo when `skipMintCheck && mintDataDir == "" && mintProject == ""` (the repo-secrets path needs a target repo). See the `needConfigRepo` variable.

2. **Config repo not needed for dev-mint path**: When `mintDataDir` is set, the `WithStoreSecret` callback now calls `writePEMFile` (PEM only, no config.json), so the config repo isn't required in that case.

File: internal/cli/admin.go [hunk 3]
```diff
@@ -1548,7 +1656,22 @@ func runInstall(ctx context.Context, client forge.Client, printer *ui.Printer, o
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/cli/admin.go:1708]: Fixed. Split `storePEMToDisk` into two functions:

- `writePEMFile(dataDir, role, pemData)` — writes only the PEM file
- `storePEMToDisk(dataDir, org, role, appID, pemData)` — writes PEM + config.json atomically

The `WithStoreSecret` callback and `skipMintDispatcher.StoreAgentPEM` now use `writePEMFile` (PEM-only). The post-loop in `runInstall`/`runPerRepoInstall` continues using `storePEMToDisk` with the correct `appID`. No more intermediate config.json with empty `app_id` values.

**Comment by ggallen** [comment]
All findings addressed:

**Medium:**
- **ensureConfigRepoExists gate**: Added `needConfigRepo` guard that also triggers when `skipMintCheck && mintDataDir == "" && mintProject == ""`, so the repo-secrets path has a target repo.
- **Duplicate PEM writes**: Split into `writePEMFile` (PEM-only) and `storePEMToDisk` (PEM + config.json). `WithStoreSecret` and `StoreAgentPEM` now use `writePEMFile`; the post-loop uses `storePEMToDisk` with the correct `appID`. No more intermediate config.json with empty `app_id`.
- **Stale ADR 0033**: Updated `--skip-mint-check` description to match actual behavior.

**Low:**
- **Path traversal in devmint**: Added `mintcore.RolePattern` validation in `loadFromDisk()` + test.
- **Type duplication**: Exported `DiskConfig`/`DiskRoleConfig` from `devmint` package; `admin.go` imports them.
- **Naming convention**: Removed local `rolePattern`; `parseAgentRoles` uses `mintcore.RolePattern` directly.
- **Stale doc comment**: Fixed `runAppSetup` comment (was mentioning /v1/pem endpoint).
- **Incomplete doc**: Added `fullsend mint run` to infrastructure-reference.md.
- **Architectural coherence (mintcore.Handler reuse)**: Acknowledged as follow-up work.

File: internal/cli/admin.go [hunk 4]
```diff
@@ -1382,25 +1476,37 @@ func runAppSetup(ctx context.Context, client forge.Client, printer *ui.Printer,
...
-	// Otherwise, store in GitHub repo secrets.
+	// Store PEMs immediately after app creation. Priority:
+	// Secret Manager (GCP) > dev mint (disk) > repo secrets.
 	if pemProvisioner != nil {
 		setup = setup.WithStoreSecret(func(sctx context.Context, role, pem string) error {
 			return pemProvisioner.StoreAgentPEM(sctx, org, role, []byte(pem))
 		})
+	} else if mintDataDir != "" {
+		setup = setup.WithStoreSecret(func(sctx context.Context, role, pem string) error {
+			return writePEMFile(mintDataDir, role, []byte(pem))
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/cli/admin.go]: Fixed. Expanded `StoreSecretFunc` signature to include `appID int`, so the `WithStoreSecret` callback for `mintDataDir` now calls `storePEMToDisk` with the correct app ID — writing both the PEM file and config.json atomically in a single call. No more crash window with missing config.json entries.

File: docs/guides/infrastructure/dev-mint.md [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+```
+
+Both should return `{"status":"ok"}`.
+
+## Step 2: Install fullsend
+
+Run the install command, pointing at the mint tunnel URL:
+
+```bash
+fullsend admin install <org> \
```
[the remark this answers is not in the document]
Comment (ggallen) [reply docs/guides/infrastructure/dev-mint.md:57]: Fixed. Added `--mint-data-dir ~/.fullsend-mint` to both the install and re-run examples, and added it to the flags table with a description.

File: docs/guides/getting-started/installation.md [hunk 6]
```diff
@@ -225,17 +225,17 @@ The installer automatically provisions [Workload Identity Federation (WIF)](http
 | `--mint-region` | `us-central1` | Cloud region for the token mint function |
```
[the remark this answers is not in the document]
Comment (ggallen) [reply docs/guides/getting-started/installation.md:248]: Fixed. Added `--mint-data-dir` row to the admin install flags table in installation.md.

File: internal/devmint/devmint.go [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+}
+
+func New(opts Options) *Server {
+	s := &Server{
+		dataDir:          opts.DataDir,
+		addr:             net.JoinHostPort(opts.Bind, fmt.Sprintf("%d", opts.Port)),
+		logger:           opts.Logger,
+		githubBaseURL:    githubAPIBaseURL,
+		httpClient:       &http.Client{Timeout: 30 * time.Second},
+		insecureNoAuth:   opts.InsecureNoAuth,
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/devmint/devmint.go:102]: Documented as accepted risk for development use. Added a "Permissive workflow and role defaults" bullet to the Limitations section in dev-mint.md explaining the `*` default and noting the `--allowed-workflows` flag for tighter controls. Also added a code comment at the default assignment in devmint.go.

File: internal/devmint/tunnel.go [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+				case urlCh <- match:
+				default:
+				}
+			}
+		}
+	}()
+
+	select {
+	case url := <-urlCh:
+		return url, cleanup, nil
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/devmint/tunnel.go:74]: Fixed. StartTunnel now uses an exitCh goroutine that calls cmd.Wait() and sends to the select statement. If cloudflared exits before emitting a URL, the error is surfaced immediately rather than waiting for the 30-second timeout. cleanup was changed to only kill (not wait) to avoid a double-Wait race with the exitCh goroutine.

File: docs/guides/infrastructure/dev-mint.md [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+
+| Flag | Purpose |
+|------|---------|
+| `--mint-url` | The mint URL. Use `http://localhost:8321` for the initial install; update to the tunnel URL in Step 3. |
+| `--skip-mint-check` | Bypasses GCP mint validation; uses the standalone mint instead |
+| `--mint-data-dir` | Path to the standalone mint's data directory (same as `--data-dir` in Step 1). PEMs and config are written here during install so the mint can serve tokens. |
+| `--app-set` | Prefix for GitHub App names (e.g., `myorg` creates `myorg-fullsend`, `myorg-triage`, etc.). Use this to avoid slug conflicts with existing public apps. |
+| `--inference-project` | GCP project ID for Vertex AI. Required for agents to call Claude. |
+
+The installer will:
```
[the remark this answers is not in the document]
Comment (ggallen) [reply docs/guides/infrastructure/dev-mint.md:73]: Fixed. All --tunnel examples in dev-mint.md now include --allowed-workflows with an example path. The code already enforces --allowed-workflows when --tunnel is set; the guide was just missing it.

File: internal/cli/mint_run.go [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+Use --insecure-no-auth to disable OIDC verification (for local testing only).`,
+		Args: cobra.NoArgs,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if dataDir == "" {
+				dataDir = os.Getenv("FULLSEND_MINT_RUN_DATA_DIR")
+			}
+			if dataDir == "" {
+				return fmt.Errorf("--data-dir is required (or set FULLSEND_MINT_RUN_DATA_DIR)")
+			}
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/cli/mint_run.go:56]: Fixed. The --insecure-no-auth loopback check now handles 'localhost' as a hostname special case alongside IP addresses.

File: internal/devmint/devmint.go [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+
+	s.logger.Printf("[MINT] token request: role=%s repos=%v", req.Role, req.Repos)
+
+	jwt, err := mintcore.GenerateAppJWT(appID, pemData)
+	if err != nil {
+		s.logger.Printf("[MINT] JWT generation failed: %v", err)
+		writeJSON(w, http.StatusInternalServerError, errorResponse{Error: "failed to generate app JWT"})
+		return
+	}
+
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/devmint/devmint.go:283]: Fixed. handleToken now logs claims.Repository and claims.JobWorkflowRef when OIDC verification succeeds. The claims variable is also used for the explicit RepositoryOwner == org defense-in-depth check.

File: internal/devmint/devmint.go [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+		verifierWorkflows := s.allowedWorkflows
+		if len(verifierWorkflows) == 0 {
+			verifierWorkflows = []string{"*"}
+		}
+		s.oidcVerifier = mintcore.NewJWKSVerifier(mintcore.JWKSVerifierConfig{
+			IssuerURL:            "https://token.actions.githubusercontent.com",
+			Audience:             s.oidcAudience,
+			AllowedOrgs:          []string{s.org},
+			AllowedWorkflowFiles: verifierWorkflows,
+			PerRepoWIFRepos:      s.perRepoWIFRepos,
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/devmint/devmint.go:384]: Documented as accepted trade-off for development use. The standalone mint is short-lived and loopback-only by default. The security considerations section in dev-mint.md explicitly contrasts the wildcard default with the production mint's explicit allowlist requirement. The --tunnel path correctly enforces --allowed-workflows.

File: internal/cli/mint_run.go [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+				if !isLoopback {
+					return fmt.Errorf("--insecure-no-auth requires a loopback bind address "+
+						"(got %q); use --bind 127.0.0.1 or omit --bind", bind)
+				}
+			}
+
+			if tunnel && len(allowedWorkflows) == 0 {
+				return fmt.Errorf("--tunnel requires --allowed-workflows: a public tunnel with wildcard " +
+					"workflow access lets any GitHub Actions workflow in the org mint tokens.\n" +
+					"  To restrict to fullsend workflows: --allowed-workflows '.github/workflows/dispatch.yml'\n" +
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/cli/mint_run.go:77]: Acknowledged. OIDC verification remains active (only --insecure-no-auth disables it, which is blocked for non-loopback bind addresses). The wildcard default is documented as an accepted trade-off for local development use.

File: docs/guides/dev/cli-internals.md [hunk 14]
```diff
@@ -18,7 +18,8 @@ fullsend
 │   ├── deploy                               # Deploy/update mint Cloud Function
 │   ├── enroll       <org|owner/repo>        # Register org/repo in mint
 │   ├── unenroll     <org|owner/repo>        # Remove org/repo from mint
-│   └── status       [org]                   # Inspect mint state and PEM health
+│   ├── status       [org]                   # Inspect mint state and PEM health
+│   └── run                                  # Run standalone token mint (no GCP required)
```
Comment (ifireball) [inline docs/guides/dev/cli-internals.md:22]: I kind of assumed we would end up with a production-worthy stand-alone version of the mint, but dev-mint.md lists many limitations that really indicate this is only for dev use, so I wonder if we should indicate this with the command name, because I do want to ultimately reach a point where we have a command that runs a production mint.

Or perhaps we need to aim toward this becoming a prod-ready deployment, file issues for the gaps, and print a loud warning at statup while it isn't.

File: docs/ADRs/0043-skip-mint-check-updated-semantics.md [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+## Decision
+
+`--skip-mint-check` now means: **skip GCP mint provisioning** (Cloud
+Function deployment, Workload Identity Federation setup, Secret Manager
+provisioning). It no longer implies skipping GitHub App creation.
+
+Specifically:
+
+- GitHub Apps **are still created** when `--skip-app-setup` is not set.
+- PEM storage is determined by flag combination, evaluated in order:
```
Comment (ifireball) [inline docs/ADRs/0043-skip-mint-check-updated-semantics.md:55]: I wonder if we should introduce variables like `--mint-platform=gcp|local` and `--mint-secret-store=gcp|file` that then require other variables as needed. This implicit emergence of the resulting architecture from the variables being set is confusing and will get more confusing when we support other deployment options such as Cloudflare and K8s. 

I'm not sure if we should have a simple variable that select the platform and implies all other settings like secret loading backend and OIDC verification backend or wither we should have different setting to let the user select different combinations (though not all combinationsmight be possible or make sense)

File: docs/guides/dev/cli-internals.md (hunk 14 above)
Comment (ggallen) [reply docs/guides/dev/cli-internals.md:22]: My goal was a mint that could be standalone and allow someone to easily install and evaluate fullsend, hence all the limitation.  We can expand that if necessary. Let me know what you think.

Comment (ifireball) [reply docs/guides/dev/cli-internals.md:22]: I want to make it so the command set under `fullsend mint` is good for setting up and managing all kinds of mint architectures, from GCP function, to Cloudflare worker to k8s deployment to local , but its hard for me to figure out what this needs to look like. I'm trying to see how to make the right choices now so we don't need a messy user-visible change later

**Comment by rh-hemartin** [comment]
@ggallen how does this relates to your standalone mint PR?

**Comment by rh-hemartin** [comment]
Does https://github.com/fullsend-ai/fullsend/pull/2537 supersedes this?

**Comment by rh-hemartin** [comment]
Does #2537 supersedes this?

**Comment by ggallen** [comment]
Superseded by #2537.
